certbot/compat/filesystem.py

```diff
@@ -0,0 +1,160 @@
+"""Compat module to handle files security on Windows and Linux"""
+from __future__ import absolute_import
+
+import os  # pylint: disable=os-module-forbidden
+import stat
+
+try:
+    import ntsecuritycon  # pylint: disable=import-error
+    import win32security  # pylint: disable=import-error
+except ImportError:
+    POSIX_MODE = True
+else:
+    POSIX_MODE = False
+
+
+def chmod(file_path, mode):
+    # type: (str, int) -> None
+    """
+    Apply a POSIX mode on given file_path:
+        * for Linux, the POSIX mode will be directly applied using chmod,
+        * for Windows, the POSIX mode will be translated into a Windows DACL that make sense for
+          Certbot context, and applied to the file using kernel calls.
+
+    The definition of the Windows DACL that correspond to a POSIX mode, in the context of Certbot,
+    is explained at https://github.com/certbot/certbot/issues/6356 and is implemented by the
+    method _generate_windows_flags().
+
+    :param str file_path: Path of the file
+    :param int mode: POSIX mode to apply
+    """
+    if POSIX_MODE:
+        os.chmod(file_path, mode)
+    else:
+        _apply_win_mode(file_path, mode)
+
+
+def _apply_win_mode(file_path, mode):
+    # Resolve symbolic links
+    if os.path.islink(file_path):
+        link_path = file_path
+        file_path = os.readlink(file_path)
+        if not os.path.isabs(file_path):
```

Comment:
Well ... I would not say that it is broken, but it is not functional in this specific context ...

Typical failure if `os.path.abspath` is this function:
https://github.com/certbot/certbot/blob/0a8005ec10cf16942a20ee9eec4ac7965a2ad8cc/certbot/tests/storage_test.py#L126-L136

The problem is that here, we have links defining a relative path target.

Let's split this up.

Supposing we are in path `C:\Tmp\dir1`. This method creates for instance, on path `C:\Tmp\dir1\live\test\cert.pem` a link pointing to the relative path `..\..\archive\test\cert1.pem`. This is a perfectly valid link, both on Windows and Linux. Target will be resolved relatively to the path of the link.

However, in `_apply_mode`, we receive the path `C:\Tmp\dir1\live\test\cert.pem`. Python sees that it is a link, and so call `os.readlink`. What `os.readlink` returns, is the target path registered in the link, so ... `..\..\archive\test\cert1.pem`. But if we call know `os.path.abspath` on this relative path, it will be resolved against the current work dir, that can be anything (like a call to `os.chdir` could have been done in the mean time). So the realpath may be incorrectly resolved to the wrong path.

It is the case here, `os.path.abspath` will return `C:\archive\test\cert1.pem`.

What you want, is to resolve the relative path against the absolute path of the link. This is what the condition under `if not os.path.isabs(file_path)`, obviously skipping this logic if the registered target path in the link is already absolute.

Reply:
Note that it is also what the official documentation advice us to do: https://docs.python.org/3/library/os.html#os.readlink

Reply:
Nice catch! That makes sense to me.